api/prometheus/v1/api.go

```diff
@@ -17,17 +17,83 @@ package v1
 
 import (
 	"context"
-	"encoding/json"
 	"errors"
 	"fmt"
 	"net/http"
 	"strconv"
 	"time"
+	"unsafe"
+
+	json "github.com/json-iterator/go"
 
 	"github.com/prometheus/client_golang/api"
 	"github.com/prometheus/common/model"
 )
 
+func init() {
+	json.RegisterTypeEncoderFunc("model.SamplePair", marshalPointJSON, marshalPointJSONIsEmpty)
+	json.RegisterTypeDecoderFunc("model.SamplePair", unMarshalPointJSON)
+}
+
+func unMarshalPointJSON(ptr unsafe.Pointer, iter *json.Iterator) {
+	p := *((*model.SamplePair)(ptr))
+	if !iter.ReadArray() {
+		iter.ReportError("unmarshal model.SamplePair", "value must be an array")
+		return
+	}
+	p.Timestamp = model.Time(iter.ReadFloat64())
+	if !iter.ReadArray() {
+		iter.ReportError("unmarshal model.SamplePair", "array must be of len 2")
+		return
+	}
+
+	f, err := strconv.ParseFloat(iter.ReadString(), 64)
+	if err != nil {
+		iter.ReportError("unmarshal model.SamplePair", err.Error())
+		return
+	}
+	p.Value = model.SampleValue(f)
+
+	if iter.ReadArray() {
+		iter.ReportError("unmarshal model.SamplePair", "array must be of len 2")
```

Comment:
I don't understand what you mean by this comment.
Maybe something like
```suggestion
		iter.ReportError("unmarshal model.SamplePair", "array has more elements than expected")
```

Reply:
The value array must have exactly 2 elements (1) timestamp and (2) value

Reply:
yes I got that part, but in this particular case if `iter.ReadArray()` is true it means that that the parsing found more elements so the correct error should be:
"array has more elements than expected" , right?